[Windows Build Scripts] - Update toolchain and dependencies versions.

Justaphf opened the pull request:
This updates the native Windows MinGW build scripts based on the work done by @CCGLLC-knc in #1219 to update the mingw build instructions.  Does not exactly follow the updated steps as there were some typos and I ran into trouble with the 32-bit Boost build of v1.66 so updated to v1.68 instead.

Summary of version updates and configuration changes necessary:
* Tool chain (both 32 & 64 bit) updated to 7.3.0, v5, rev0
* Update hexdump utility to latest master build (only used if building in tests)
* Update OpenSSL to v1.0.2o
  * Configuration needed to be changed to call `make depend` before calling `make`
* Update libpng to v1.6.36
  * Required applying a patch to fix line ending handling prior to building
* Update qrencode to v4.0.2
* Update Boost to v1.68.0
  * Configuration now requires specifying `address-model` when building for 32-bit.  Not sure if this is because of the toolchain or because of Boost itself (v1.61 with the v4.9.2 toolchain didn't require this)
* Update QT to v5.7.1
  * Required applying a patch for static library builds
  * Configuration required changing to use `qt-zlib`, `qt-libpng`, and `qt-libjpeg`.  Previously the v5.3.2 Qt configuration used `system-zlib`, `system-libpng` and `no-libjpeg`.
  * Configuration also now requires explicit specification of include and lib folders for the OpenSSL path
  * Bitcoin configuration required additionally specifying the Qt libs location in addition to the existing `with-qt-libdir`

I have tested full download and build from scratch of all dependencies and bitcoin binaries in both 32-bit and 64-bit modes.  Ran each of the 32-bit and 64-bit builds for about a day each without seeing issues with normal operations.

Justaphf committed:
Update hexdump utility source to master as of 2018-12-21

A new release was tagged on 2018-12-15, however, with gcc 7.3.0 using the -Werror option causes a build failure.  This is fixed in the subsequent patch from 2018-12-21, but has not been tagged as a release yet.

Justaphf committed:
Update OpenSSL dependency to v1.0.2o

Justaphf committed:
Include patch utility in base MinGW installation

Justaphf committed:
Update libpng dependency to v1.6.36

Justaphf committed:
Update qrencode dependency to v4.0.2

Justaphf committed:
Update MinGW build toolchain to gcc v7.3.0, v5, rev0

Justaphf committed:
Update Boost dependency to v1.68.0

Justaphf committed:
Update Qt dependency to v5.7.1

Justaphf committed:
Add bitcoin-miner.exe to post-build processing

Justaphf committed:
Change default path for git install location

sickpig reviewed (commented):
@Justaphf thanks for doing that. Quickly went through the changes and all seems sound tome. Didn't test it thou since I don't have a Win machine to test it on.

ptschip commented:
@Justaphf can't seem to get libpng compiled...i think it's not finding the patch to apply

ptschip commented:
here's some output from the libpng compile:

make[1]: Entering directory `/D/deps/x86/libpng-1.6.36'
depbase=`echo contrib/tools/pngfix.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
        gcc -DHAVE_CONFIG_H -I.     -g -O2 -MT contrib/tools/pngfix.o -MD -MP -MF $depbase.Tpo -c -o contrib/tools/pngfix.o contrib/tools/pngfix.c &&\
        mv -f $depbase.Tpo $depbase.Po
In file included from contrib/tools/../../png.h:330:0,
                 from contrib/tools/pngfix.c:44:
contrib/tools/../../pnglibconf.h:206:54: error: expected identifier or '(' before '-' token

ptschip commented:
Ok I verified the patch was applied but still no luck on compiling.

ptschip commented:
@Justaphf all right, nvm, just me...i didn't set the github path correct , had an extra '"' in there ...building now.

ptschip commented:
@Justaphf sorry for all the noise but i did find one nit:   in clean-bitcoin.sh there's still a reference to boost 1.61

export BOOST_ROOT="$PATH_DEPS/boost_1_61_0"

ptschip commented:
@Justaphf one other nit...my builds can never seem to find my python installation so I've in the past hacked this in...I just add %PATH% to the end.  If it makes sense it would be great to add this in to the code as well...

REM Add MSYS and MinGW bin directories to the start of path so commands are available
set "PATH=%MSYS_BIN%;%MINGW_BIN%;%BASE_PATH%;%PATH%"

Justaphf commented:
@ptschip Thanks for reviewing & testing.
I've corrected the clean script bug to point to the correct version of boost.

For the python issue, what version(s) of python do you have installed?  If you have 3.x installed, it's better to use the python version selector `py.exe` than directly specifying the 2.x and 3.x separately.  One thing to note is that the python location(s) specified in the `SET_ENV_VARS.bat` are only actually used by `install-utils.sh`, which in turn is only called by `config-mingw.bat`.  If you make changes to those settings but only run `rebuild-bitcoin.bat` it won't affect the executable scripts that wrap the calls to the Windows executables that are aliased to the `python`, `python2`, and `python3` commands.

I think if I just always call the `install-utils.sh` script inside the `rebuild-bitcoin.bat` script that may take care of it.  I'll work on implementing and testing that and if it seems OK on my machine I'll push it.

The reason I originally dropped having the original Windows %PATH% specified was because of issues I was having on my test machines.  Initially my %PATH% was too long so I got errors because of that, then when I manually trimmed it down so it wasn't too long, the install of Anaconda3 (automatically installed by VS2017 with the R component enabled) was redirecting some calls when building the Qt dependency causing it to fail to build.  This was even with the path variable specified at the end.  So I trimmed it down to only the paths that were needed for the build to work on my machine.

It may be that the newer version of Qt doesn't suffer this issue, but I haven't tested reverting back to the full %PATH% yet.  If running `install-utils.sh` as part of the `rebuild-bitcoin.bat` script doesn't fix it, I'll try that.

Also note, I never got the python tests to run natively on my Windows system, only the `make check` tests, and of those, only some of them passed.

Justaphf committed:
Call install-utilities.sh from rebuild-bitcoin.bat

This will ensure that if the user modifies the paths to git or python in SET_ENV_VARS.bat, the executable wrappers are updated accordingly.
Without this change, the user needs to rerun config-mingw.bat to update the utility wrappers.

ptschip reviewed (approved):
Looks good...

ptschip commented:
@gandrewstone tested this out and works...also we need it for c++14

Justaphf commented:
@ptschip I added in the call to update the git and python paths every time the rebuild script is run.  This way users don't need to rerun `config-mingw.bat` just to change the path to those.  I also discovered an issue for 64-bit with tests enabled.  Basically I need to explicitly specify the boost architecture to 64-bit like I do for the 32-bit builds.  This corrects the linking for the boost test framework.

gandrewstone commented:
thanks for all your work on the windows platform!
